drafts/ERROR_HANDLING.md

```diff
@@ -0,0 +1,138 @@
+
+
+### Application Level
+Application level errors have codes in different ranges. 
+* 1000 - 1499 range for field validation / input errors
+* 1501 - 1999 range for business logic violations
+* 9999 is for unhandled and unexpected situations
+* Capabilities may use overlapping error codes but will be uniquely namespaced. See the capability's documentation for more details.
+
+## Complete Error Overview
+### HTTP 400
+The errors outlined below are believed to be errors that one will want to act / alert on as they impact the ability to continue the logical call-flow. 
+They MUST be accompanied by an HTTP 400 status code
+
+**Generic**
+- `ErrorCode` 1499: Something is wrong with the input. `errorMessage` should indicate what exactly.
+- `ErrorCode` 1498: Unable to create entity. `errorMessage` should indicate what exactly. Use when unable to process a POST request and no more explicit error code is available.
+- `ErrorCode` 1497: Unable to update entity. `errorMessage` should indicate what exactly. Use when unable to process a PUT/PATCH request and no more explicit error code is available.
+
+**Supplier** 
+No errors identified
+
+**Products**
+- `ErrorCode` 1001: `supplierId` not found / invalid
+- `ErrorCode` 1002: `productId` not found / invalid
+- `ErrorCode` 1003: `optionId` not found / invalid
+
+**Availability**
+- `ErrorCode` 1001: `supplierId` not found / invalid
+- `ErrorCode` 1002: `productId` not found / invalid
+- `ErrorCode` 1003: `optionId` not found / invalid
+- `ErrorCode` 1004: `availabilityId` not found / invalid
+
+**Reservations**
+- `ErrorCode` 1001: `supplierId` not found / invalid
+- `ErrorCode` 1002: `productId` not found / invalid
+- `ErrorCode` 1003: `optionId` not found / invalid
+- `ErrorCode` 1004: `availabilityId` not found / invalid
+- `ErrorCode` 1005: `uuid` of reservation/booking not found / invalid
+- `ErrorCode` 1501: Reservation expired 
+- `ErrorCode` 1502: Hold period too long 
+- `ErrorCode` 1503: Cannot extend hold period 
+- `ErrorCode` 1504: Reservation ID not found
+- `ErrorCode` 1505: No more availability
+
+**Bookings**
+- `ErrorCode` 1001: `supplierId` not found / invalid
+- `ErrorCode` 1005: `uuid` of reservation/booking not found / invalid
+- `ErrorCode` 1501: Reservation expired 
+
+**Cancellations**
+- `ErrorCode` 1005: `uuid` of reservation/booking not found / invalid
+- `ErrorCode` 1506: Cancellation not allowed
+
+### HTTP 500
+System errors that occur in the software that embeds the ICF may cause erratic behavior or cause hard-to-map errors. Those should be thrown with an HTTP 500 status code.
+```javascript
+[{
+	"errorCode": 9999,
+	"errorMessage": "System.outOfMemoryException thrown in /some/file",
+	"fieldName": null
+}]
+```
+### Other guidelines
+- An error during a "create" action such as placing a booking should be a guarantee that the booking was not finalized and shall not be charged. **ONE FOR THE SLA MAYBE?**
+
+
+
+
+## NOT FOR PUBLISHING
```

Comment:
Should this section be removed or cleaned up?

Reply:
Yeah, I'll move that.
I felt it needed to be stored somewhere as it's a guideline for future content writing / decision making so I parked it there first.